Move the game off of surfaces, and use the render system instead

AllyTally opened the pull request:
## Changes:

This is quite a big PR.

Ever since VVVVVV was initially ported to C++ in 2.0, it has used surfaces from SDL. The downside is, that's all software rendering. This PR moves most things off of surfaces, and all into GPU, by using textures and SDL_Renderer.

Pixel-perfect collision has been kept by keeping a copy of sprites as surfaces. There's plans for pixel-perfect collision to use masks instead of reading pixel data directly, but that's out of scope for this PR.

Everything has been ported over, and from a bit of playtesting, I see no visual differences between the old system and the new one. This PR should help make the code more understandable and easier to mess with, while also providing a fair framerate boost, cutting down on CPU, and it maybe even cuts down on memory a bit.

Also as a quick note, `graphics.reloadresources()` in `main` now gets called _after_ the renderer is created, because textures cannot be created without a renderer.

## Legal Stuff:

By submitting this pull request, I confirm that...

- [x] My changes may be used in a future commercial release of VVVVVV
- [x] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

flibitijibibo commented:
If we render the full game frame to a target we should be able to support linear as well as analogue mode - it would mean implementing our own scale logic but that's not too hard to pull from SDL's logical size work. Analogue mode would just lock the render target and do the operation on the CPU; not great but probably still faster than the all-software path we have now.

(As for VRAM, it probably puts everything in host memory at first unless it's Vulkan/D3D12 where we can explicitly put everything in device-local memory ourselves. This is likely what we'll do with the SDL_gpu backend for SDL3.)

InfoTeddy reviewed (changes requested):
Screen shake does not render properly with scaling.

https://user-images.githubusercontent.com/59748578/211213193-8e71941e-e5cd-4152-9a2c-9e62376041d5.mp4

AllyTally commented:
> If we render the full game frame to a target we should be able to support linear as well as analogue mode - it would mean implementing our own scale logic but that's not too hard to pull from SDL's logical size work.
The game does have a target texture for the entire thing, `graphics.gameTexture`, do you have an example of how this'd work? I mostly had problems figuring out how to make it toggleable at runtime.

flibitijibibo commented:
I think there's an API to set the texture address mode but the old way to do it was to use SDL_HINT_RENDER_SCALE_QUALITY to cause it to filter in linear when doing the final upscale, but it may have a different method available that we didn't have in 2014...

AllyTally commented:
Interesting, the SDL_HINT_RENDER_SCALE_QUALITY thing was only working yesterday when it happened before the renderer was created. I might have been doing something wrong though -- I'll look into this tomorrow

InfoTeddy reviewed (changes requested):
Valgrind reports several memory errors upon exiting the game (see the attachment).

* [valgrind_errors.txt](https://github.com/TerryCavanagh/VVVVVV/files/10369881/valgrind_errors.txt)

This seems to be because the renderer for these textures has already been destroyed at this point, thus invalidating those textures.

AllyTally commented:
I've reimplemented linear filtering and analogue mode, fixed the shaking bug mentioned and removed unused functions and variables (no reason to keep the screen surface and old screen texture around if they're not used anymore!)

Dav999-v commented:
I used my old fridge to benchmark before and after this PR. :)

![I honestly wasn't expecting more than 30fps on this thing](https://user-images.githubusercontent.com/44736680/211418349-2ed160ef-8f4c-4676-b50c-28263f93c8c0.jpg)
![Specs](https://user-images.githubusercontent.com/44736680/211418485-643fc8a5-db3f-494e-b8a5-2c661892efbf.png)

First the CPU usage, which I just eyeballed by looking at task manager for a minute each time:

|  | 30fps mode | over 30fps mode |
| --- | --- | --- |
| upstream/master | 30-38% (peak 48) | 81-89% (peak 91) |
| AllyTally/renderer-neo | 13-20% | 47-70% (peak 78 but it's more wild) |

Then the frames per second, all done with over 30fps mode enabled:

|  | #Frames | Time | Avg fps | Min fps | Max fps |
| --- | --- | --- | --- | --- | --- |
| upstream/master, vsync off | 2986 | 60000ms | 49.767 | 50 | 53 |
| AllyTally/renderer-neo, vsync off | 3340 | 60000ms | 55.667 | 53 | 58 |
| upstream/master, vsync on | 2812 | 60000ms | 46.867 | 45 | 48 |
| AllyTally/renderer-neo, vsync on | 3293 | 60000ms | 54.883 | 53 | 57 |

![vsync off](https://user-images.githubusercontent.com/44736680/211432844-434dd6bc-8ba2-4ad0-9cc5-ca0e52ae19a7.png)
![vsync on](https://user-images.githubusercontent.com/44736680/211432894-fd003710-6ca1-4b8b-a9b6-1055f02a9b91.png)

I also did the same FPS benchmarks on my newest computer (with vsync off) - here's a neofetch, but I used Windows 11 for the tests:
![neofetch](https://user-images.githubusercontent.com/44736680/211433006-c296bf42-ef70-4113-a366-809763cc14b7.png)

|  | #Frames | Time | Avg fps | Min fps | Max fps |
| --- | --- | --- | --- | --- | --- |
| upstream/master | 95555 | 60000ms | 1592.583 | 1555 | 1602 |
| AllyTally/renderer-neo | 119850 | 60000ms | 1997.500 | 1841 | 2018 |

![Ryzen 5](https://user-images.githubusercontent.com/44736680/211433613-f6a30fb6-3c3b-4c27-9144-951b46848059.png)

So it's definitely a noticeable improvement! And as far as my knowledge goes, these are two systems with integrated graphics, so using the GPU instead of the CPU on these isn't as enormous of an improvement as it could be with a dedicated GPU.

win8linux commented:
Just to be clear, will the option of software rendering still be available for platforms that do not have GPU acceleration? I am maintaining the Haiku port and packages associated with it, a platform with very limited GPU acceleration.

InfoTeddy commented:
> Just to be clear, will the option of software rendering still be available for platforms that do not have GPU acceleration? I am maintaining the Haiku port and packages associated with it, a platform with very limited GPU acceleration.

You will always be able to use the software fallback renderer by passing `-renderer software` in command-line options.

Dav999-v commented:
Since I heard caching improved even further, I did some more fridge benchmarks (renderer-09 is the one from 9 Jan, renderer-20 is today's build):

|  | #Frames | Time | Avg fps | Min fps | Max fps |
| --- | --- | --- | --- | --- | --- |
| renderer-20, vsync off | 3635 | 60000ms | 60.583 | 56 | 65 |
| renderer-20, vsync on | 3533 | 60000ms | 58.883 | 56 | 61 |

![vsync off](https://user-images.githubusercontent.com/44736680/213821571-f08018f0-54c0-47e3-b44f-79dba71d8af2.png)
![vsync on](https://user-images.githubusercontent.com/44736680/213821609-93af22ef-390a-450b-887b-50977b202bc4.png)

(`upstream` and `renderer-09` in the graph are reused from last time, I did re-benchmark the old one just in case and it basically didn't change from last time:)

|  | #Frames | Time | Avg fps | Min fps | Max fps |
| --- | --- | --- | --- | --- | --- |
| renderer-09, vsync off | 3343 | 60000ms | 55.717 | 53 | 58 |
| renderer-09, vsync on | 3344 | 60000ms | 55.733 | 54 | 58 |

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -160,101 +157,41 @@ void Graphics::init(void)
-    SDL_SetSurfaceBlendMode(menuoffbuffer, SDL_BLENDMODE_NONE);
-
-    #undef CREATE_SURFACE
+    gameTexture       = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    gameplayTexture   = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    menuTexture       = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    ghostTexture      = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    tempTexture       = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    foregroundTexture = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS);
+    backgroundTexture = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS + 16, SCREEN_HEIGHT_PIXELS + 16);
+    towerbg.texture   = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS + 16, SCREEN_HEIGHT_PIXELS + 16);
+    titlebg.texture   = SDL_CreateTexture(gameScreen.m_renderer, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, SCREEN_WIDTH_PIXELS + 16, SCREEN_HEIGHT_PIXELS + 16);
```
The previous code was less copy-pasted and I'd like to keep it that way. Also, in general, I don't like alignment spacing stuff because it necessitates updating all other lines to keep the alignment in sync even if only one line is changed.
```suggestion
#define CREATE_TEXTURE_WITH_DIMENSIONS(w, h) \
    SDL_CreateTexture( \
        gameScreen.m_renderer, \
        SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_TARGET, \
        (w), (h) \
    )
#define CREATE_TEXTURE \
    CREATE_TEXTURE_WITH_DIMENSIONS(SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS)
#define CREATE_SCROLL_TEXTURE \
    CREATE_TEXTURE_WITH_DIMENSIONS(SCREEN_WIDTH_PIXELS + 16, SCREEN_WIDTH_PIXELS + 16)

    gameTexture = CREATE_TEXTURE;
    gameplayTexture = CREATE_TEXTURE;
    menuTexture = CREATE_TEXTURE;
    ghostTexture = CREATE_TEXTURE;
    tempTexture = CREATE_TEXTURE;
    foregroundTexture = CREATE_TEXTURE;
    backgroundTexture = CREATE_SCROLL_TEXTURE;
    towerbg.texture = CREATE_SCROLL_TEXTURE;
    titlebg.texture = CREATE_SCROLL_TEXTURE;
    
#undef CREATE_SCROLL_TEXTURE
#undef CREATE_TEXTURE
#undef CREATE_TEXTURE_WITH_DIMENSIONS
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -2252,24 +2327,26 @@ void Graphics::drawentity(const int i, const int yoff)
     case 4:    // Small pickups
-        drawhuetile(xp, yp - yoff, obj.entities[i].tile, obj.entities[i].realcol);
+    {
+        SDL_Color color = obj.entities[i].realcol;
```
```suggestion
        const SDL_Color color = obj.entities[i].realcol;
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -2629,27 +2675,35 @@ void Graphics::drawbackground( int t )
+    {
+        clear();
+
+        int offset = (int)lerp(0, -3);
+        SDL_Rect srcRect = { 8 + offset, 0, 320, 240 };
```
```suggestion
        const int offset = (int) lerp(0, -3);
        const SDL_Rect srcRect = {8 + offset, 0, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -2629,27 +2675,35 @@ void Graphics::drawbackground( int t )
+    {
+        clear();
+
+        int offset = (int)lerp(0, -3);
+        SDL_Rect srcRect = { 0, 8 + offset, 320, 240 };
```
```suggestion
        const int offset = (int) lerp(0, -3);
        const SDL_Rect srcRect = {0, 8 + offset, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -2695,21 +2749,21 @@ void Graphics::drawbackground( int t )
         {
             const int temp = (i * 16) + backoffset;
-            const SDL_Rect warprect = {160 - temp, 120 - temp, temp * 2, temp * 2};
+            SDL_Rect warprect = {160 - temp, 120 - temp, temp * 2, temp * 2};
```
```suggestion
            const SDL_Rect warprect = {160 - temp, 120 - temp, temp * 2, temp * 2};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -3026,20 +3101,27 @@ void Graphics::drawtowerspikes(void)
-    SDL_BlitSurface(bg_obj.buffer_lerp, &towerbuffer_rect, backBuffer, NULL);
+    clear();
+
+    int offset = (int)lerp(0, -bg_obj.bscroll);
+
+    SDL_Rect srcRect = {0, 8 + offset, 320, 240};
```
```suggestion
    const int offset = (int) lerp(0, -bg_obj.bscroll);
    const SDL_Rect srcRect = {0, 8 + offset, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -3251,31 +3334,17 @@ SDL_Color Graphics::getcol( int t )
-    tpoint.x = x;
-    tpoint.y = y;
 
+    SDL_Rect offsetRect = { 0, (int)lerp(oldmenuoffset, menuoffset), SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS };
```
```suggestion
    const int offset = (int) lerp(oldmenuoffset, menuoffset);
    const SDL_Rect offsetRect = {0, offset, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -3441,15 +3510,30 @@ int Graphics::crewcolour(const int t)
+    clear();
 
-    ClearSurface(backBuffer);
+    SDL_Rect shake = { screenshake_x, screenshake_y, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS };
```
```suggestion
    const SDL_Rect shake = {screenshake_x, screenshake_y, SCREEN_WIDTH_PIXELS, SCREEN_HEIGHT_PIXELS};
```

InfoTeddy commented on desktop_version/src/Graphics.cpp:
```diff
@@ -3689,27 +3750,23 @@ bool Graphics::reloadresources(void)
-    images.push_back(grphx.im_image10);
-    images.push_back(grphx.im_image11);
-    images.push_back(grphx.im_image12);
+    images[IMAGE_LEVELCOMPLETE]     = grphx.im_image0;
+    images[IMAGE_MINIMAP]           = grphx.im_image1;
+    images[IMAGE_COVERED]           = grphx.im_image2;
+    images[IMAGE_ELEPHANT]          = grphx.im_image3;
+    images[IMAGE_GAMECOMPLETE]      = grphx.im_image4;
+    images[IMAGE_FLIPLEVELCOMPLETE] = grphx.im_image5;
+    images[IMAGE_FLIPGAMECOMPLETE]  = grphx.im_image6;
+
+    images[IMAGE_SITE]              = grphx.im_image7;
+    images[IMAGE_SITE2]             = grphx.im_image8;
+    images[IMAGE_SITE3]             = grphx.im_image9;
+    images[IMAGE_ENDING]            = grphx.im_image10;
+    images[IMAGE_SITE4]             = grphx.im_image11;
+    images[IMAGE_CUSTOMMINIMAP]     = grphx.im_image12;
```
```suggestion
    images[IMAGE_LEVELCOMPLETE] = grphx.im_image0;
    images[IMAGE_MINIMAP] = grphx.im_image1;
    images[IMAGE_COVERED] = grphx.im_image2;
    images[IMAGE_ELEPHANT] = grphx.im_image3;
    images[IMAGE_GAMECOMPLETE] = grphx.im_image4;
    images[IMAGE_FLIPLEVELCOMPLETE] = grphx.im_image5;
    images[IMAGE_FLIPGAMECOMPLETE] = grphx.im_image6;

    images[IMAGE_SITE] = grphx.im_image7;
    images[IMAGE_SITE2] = grphx.im_image8;
    images[IMAGE_SITE3] = grphx.im_image9;
    images[IMAGE_ENDING] = grphx.im_image10;
    images[IMAGE_SITE4] = grphx.im_image11;
    images[IMAGE_CUSTOMMINIMAP] = grphx.im_image12;
```

InfoTeddy commented on desktop_version/src/Graphics.h:
```diff
@@ -134,17 +146,63 @@ class Graphics
+    int clear(int r, int g, int b, int a);
+    int clear();
+
+    int copy_texture(SDL_Texture* texture, SDL_Rect* src, SDL_Rect* dest);
+    int copy_texture(SDL_Texture* texture, SDL_Rect* src, SDL_Rect* dest, double angle, SDL_Point* center, SDL_RendererFlip flip);
```
```suggestion
    int copy_texture(SDL_Texture* texture, const SDL_Rect* src, const SDL_Rect* dest);
    int copy_texture(SDL_Texture* texture, const SDL_Rect* src, const SDL_Rect* dest, double angle, const SDL_Point* center, SDL_RendererFlip flip);
```

InfoTeddy commented on desktop_version/src/Graphics.h:
```diff
@@ -134,17 +146,63 @@ class Graphics
+    int set_color(int r, int g, int b);
+    int set_color(SDL_Color color);
+
+    int fill_rect(SDL_Rect* rect, const int r, const int g, const int b, const int a);
+    int fill_rect(const int x, const int y, const int w, const int h, const int r, const int g, const int b, const int a);
+    int fill_rect(const int x, const int y, const int w, const int h, const int r, const int g, const int b);
+    int fill_rect(const int r, const int g, const int b, const int a);
+    int fill_rect(SDL_Rect* rect, const int r, const int g, const int b);
+    int fill_rect(const int r, const int g, const int b);
+    int fill_rect(SDL_Rect* rect, const SDL_Color color);
+    int fill_rect(const int x, const int y, const int w, const int h, const SDL_Color color);
+    int fill_rect(const SDL_Color color);
```
In header files, `const` does not matter for non-pointer arguments.
```suggestion
    int fill_rect(const SDL_Rect* rect, int r, int g, int b, int a);
    int fill_rect(int x, int y, int w, int h, int r, int g, int b, int a);
    int fill_rect(int x, int y, int w, int h, int r, int g, int b);
    int fill_rect(int r, int g, int b, int a);
    int fill_rect(const SDL_Rect* rect, int r, int g, int b);
    int fill_rect(int r, int g, int b);
    int fill_rect(const SDL_Rect* rect, SDL_Color color);
    int fill_rect(int x, int y, int w, int h, SDL_Color color);
    int fill_rect(SDL_Color color);
```

InfoTeddy commented on desktop_version/src/Graphics.h:
```diff
@@ -21,6 +21,23 @@ enum FadeBars
+    IMAGE_SITE3,
+    IMAGE_ENDING,
+    IMAGE_SITE4,
+    IMAGE_CUSTOMMINIMAP
```
```suggestion
    IMAGE_CUSTOMMINIMAP,
    NUM_IMAGES
```

InfoTeddy commented on desktop_version/src/Graphics.h:
```diff
@@ -261,28 +319,22 @@ class Graphics
-    std::vector <SDL_Surface*> flipsprites;
-    std::vector <SDL_Surface*> bfont;
-    std::vector <SDL_Surface*> flipbfont;
+    SDL_Texture* images[13];
```
```suggestion
    SDL_Texture* images[NUM_IMAGES];
```

InfoTeddy commented on desktop_version/src/Screen.cpp:
```diff
@@ -354,20 +284,49 @@ void Screen::toggleLinearFilter(void)
 {
     vsync = !vsync;
-    SDL_RenderSetVSync(m_renderer, (int) vsync);
+    SDL_RenderSetVSync(m_renderer, (int)vsync);
```
```suggestion
    SDL_RenderSetVSync(m_renderer, (int) vsync);
```

InfoTeddy commented on desktop_version/src/GraphicsResources.cpp:
```diff
@@ -47,69 +44,258 @@ SDL_Surface* LoadImage(const char *filename)
+            {
+                Uint8 r, g, b, a;
+                SDL_GetRGBA(((Uint32*)loadedImage->pixels)[i], loadedImage->format, &r, &g, &b, &a);
+                ((Uint32*)loadedImage->pixels)[i] = SDL_MapRGBA(loadedImage->format, 255, 255, 255, a);
+            }
+            SDL_UnlockSurface(loadedImage);
+            break;
+        case TEX_GRAYSCALE:
+            SDL_LockSurface(loadedImage);
+            for (int i = 0; i < loadedImage->w * loadedImage->h; i++)
+            {
+                Uint8 r, g, b, a;
+                SDL_GetRGBA(((Uint32*)loadedImage->pixels)[i], loadedImage->format, &r, &g, &b, &a);
+
+                // Magic numbers used for grayscaling (eyes percieve certain colors brighter than others)
+                r *= 0.299;
+                g *= 0.587;
+                b *= 0.114;
+
+                const double gray = SDL_floor(r + g + b + 0.5);
+
+                ((Uint32*)loadedImage->pixels)[i] = SDL_MapRGBA(loadedImage->format, gray, gray, gray, a);
+            }
+            SDL_UnlockSurface(loadedImage);
+            break;
+        default:
+            break;
+        }
+
+        //Create texture from surface pixels
+        SDL_Texture* texture = SDL_CreateTextureFromSurface(gameScreen.m_renderer, loadedImage);
+        if (texture == NULL)
+        {
+            vlog_error("Failed creating texture: %s. SDL error: %s\n", filename, SDL_GetError());
+        }
+
+        return texture;
+    }
+
+    return NULL;
```
If you're enclosing the entire function in an if-statement, then you should move it out of the if-statement and negate the conditional. Especially if the conditional is a sanity check like making sure the argument isn't `NULL`.

Secondly, you should use the `DrawPixel` and `ReadPixel` functions from `GraphicsUtil.cpp` instead of re-implementing the logic here. (`DrawPixel` will need to be un-`static`ed and exported in `GraphicsUtil.h`.)

Also:
- Put a space between a cast and the expression it's casting.
- You misspelled "perceive".
- You forgot a space between a comment starter `//` and the text of the comment.
- You ended a parenthetical in a comment with a comma ("`(which will be changed later,)`").

InfoTeddy commented on desktop_version/src/GraphicsUtil.cpp:
```diff
@@ -336,15 +144,40 @@ void UpdateFilter(void)
+        return;
+    }
+
+    int result = SDL_RenderReadPixels(gameScreen.m_renderer, NULL, 0, src->pixels, src->pitch);
```
```suggestion
    const int result = SDL_RenderReadPixels(gameScreen.m_renderer, NULL, 0, src->pixels, src->pitch);
```

InfoTeddy commented on desktop_version/src/GraphicsResources.cpp:
```diff
@@ -47,69 +44,258 @@ SDL_Surface* LoadImage(const char *filename)
+    unsigned char* data;
+    SDL_Surface* loadedImage = LoadImageRaw(filename, &data);
+
+
```
```suggestion
```

InfoTeddy commented on desktop_version/src/preloader.cpp:
```diff
@@ -114,12 +114,12 @@ void preloaderrender(void)
     //Switch to TITLEMODE (handled by preloaderrenderfixed)
   }else if (pre_transition < 5) {
-    ClearSurface(graphics.backBuffer);
+      graphics.fill_rect(0, 0, 0);
```
```suggestion
    graphics.fill_rect(0, 0, 0);
```

InfoTeddy reviewed (changes requested):
Final style pass.

In terms of testing, my Nova TAS (resynced for 2.3) still syncs. So it seems good enough to me.

We can address working around texture size hardware limits and support for very large custom tilesheets later.

I'll merge this after you fix these style issues and squash everything down to one commit (rebasing on latest `master`).

AllyTally commented:
Fixed everything mentioned, rebased and force-pushed.

InfoTeddy reviewed (changes requested):
Okay, I just found a bug - if you turn on Flip Mode from in-game options, then returning to the game will bring the menu down against a black screen. (This doesn't happen if you're turning off Flip Mode, or if you're already in Flip Mode in the first place.)

AllyTally commented:
fixed and force pushed, the bug was with screen shaking outside of gameplay

InfoTeddy reviewed (approved):
Thanks!
